### PR TITLE
Particle pusher acceleration

### PR DESCRIPTION
**Pull request by steindev**
Implements a particle pusher which accelerates particles by a constant force. This resembles the idealized situation of accelerating in the electric field of a planar capacitor.
The electric field strength and duration of acceleration can be adjusted in the `pusher.param`. The pusher was tested by initializing a `gaussianCloud` electron gas and accelerating at a field strength of `1GeV/cm` for 518fs(=2225 DT).
image

If requested, I can upload an example setup for its use.

### CODE REVIEW

**Comment by ax3l** [comment]
@steindev can you please squash and rebase against the latest `dev`? :)

**Comment by ax3l** [comment]
looks like you might need to fetch mainline first :)

File: include/picongpu/param/pusher.param [hunk 1]
```diff
@@ -28,6 +28,31 @@
 
 namespace picongpu
 {
+    struct particlePusherAccelerationParam
+    {
+        /** Define strength of constant and homogeneous acclerating
```
Comment (PrometheusPi) [inline include/picongpu/param/pusher.param]:
```suggestion
        /** Define strength of constant and homogeneous accelerating
```

File: include/picongpu/param/pusher.param [hunk 2]
```diff
@@ -28,6 +28,31 @@
...
+         * The moving window propagation direction
+         * unit: Volt / meter
+         * (1e11 V/m = 1 GV/cm)
+         */
+        static constexpr float_64 AMPLITUDEy_SI= -1.e11;
+
+        /** unit: Volt / meter */
+        static constexpr float_64 AMPLITUDEz_SI= 0.0;
+
+        /** Acceleration Duration
```
Comment (PrometheusPi) [inline include/picongpu/param/pusher.param]:
```suggestion
        /** Acceleration duration
```

File: include/picongpu/particles/pusher/particlePusherAcceleration.hpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+        using MomType = momentum::type;
+        MomType new_mom = particle[ momentum_ ];
+
+        const float_X deltaT = DELTA_T;
+
+        // normalize input SI values to
+        const float3_X eField(UnitlessParam::AMPLITUDEx, UnitlessParam::AMPLITUDEy, UnitlessParam::AMPLITUDEz);
+
+        /* Maybe refactor later to ensure a smooth and slow incease */
```
Comment (PrometheusPi) [inline include/picongpu/particles/pusher/particlePusherAcceleration.hpp]:
```suggestion
        /* ToDo: Maybe refactor later to ensure a smooth and slow increase */
```

File: include/picongpu/particles/pusher/particlePusherAcceleration.hpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    }
+
+    static pmacc::traits::StringProperty getStringProperties()
+    {
+        pmacc::traits::StringProperty propList( "name", "Acceleration" );
+        return propList;
+    }
+};
+} // namespace particlePusherBoris
```
Comment (PrometheusPi) [inline include/picongpu/particles/pusher/particlePusherAcceleration.hpp]:
```suggestion
} // namespace particlePusherAcceleration
```

File: include/picongpu/param/species.param [hunk 5]
```diff
@@ -76,6 +76,7 @@ using UsedParticleCurrentSolver = currentSolver::Esirkepov< UsedParticleShape >;
  *                                              with classical radiation reaction
  *
  * For diagnostics & modeling: ------------------------------------------------
+ * - particles::pusher::Acceleration : constant force on particles for acceleration
```
Comment (sbastrakov) [inline include/picongpu/param/species.param]: I think it is constant field, but not necessarily constant force, as the latter depends on the charge?

File: include/picongpu/particles/pusher/particlePusherAcceleration.hpp (hunk 3 above)
Comment (sbastrakov) [reply include/picongpu/particles/pusher/particlePusherAcceleration.hpp]: Sorry for nitpicking, but I don't get the meaning of this comment at all. A smooth and slow increase of what?

File: include/picongpu/particles/pusher/particlePusherAcceleration.hpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        /* Maybe refactor later to ensure a smooth and slow incease */
+        if ( currentStep * DELTA_T <= UnitlessParam::ACCELERATION_TIME )
+            new_mom += charge * eField * deltaT;
+
+        particle[ momentum_ ] = new_mom;
+
+        Velocity velocity;
+        const float3_X vel = velocity(new_mom, mass);
+
+        for(uint32_t d=0;d<simDim;++d)
```
Comment (psychocoderHPC) [inline include/picongpu/particles/pusher/particlePusherAcceleration.hpp]: missing spaces

File: include/picongpu/particles/pusher/particlePusherAcceleration.hpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        const float3_X eField(UnitlessParam::AMPLITUDEx, UnitlessParam::AMPLITUDEy, UnitlessParam::AMPLITUDEz);
+
+        /* Maybe refactor later to ensure a smooth and slow incease */
+        if ( currentStep * DELTA_T <= UnitlessParam::ACCELERATION_TIME )
+            new_mom += charge * eField * deltaT;
+
+        particle[ momentum_ ] = new_mom;
+
+        Velocity velocity;
+        const float3_X vel = velocity(new_mom, mass);
```
Comment (psychocoderHPC) [inline include/picongpu/particles/pusher/particlePusherAcceleration.hpp]: missing spaces

File: include/picongpu/particles/pusher/particlePusherAcceleration.hpp (hunk 3 above)
Comment (steindev) [reply include/picongpu/particles/pusher/particlePusherAcceleration.hpp]: I agree, the comment is too short. I state it more precisely.

File: include/picongpu/param/pusher.param [hunk 8]
```diff
@@ -28,6 +28,31 @@
...
 namespace picongpu
 {
+    struct particlePusherAccelerationParam
+    {
+        /** Define strength of constant and homogeneous accelerating
+         *  electric field in SI per dimension.
+         */
+
+        /** unit: Volt / meter */
+        static constexpr float_64 AMPLITUDEx_SI= 0.0;
```
Comment (ax3l) [inline include/picongpu/param/pusher.param]: style: spaces around `=` pls (3x in this file)

File: include/picongpu/particles/pusher/particlePusherAcceleration.hpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+};
+
+template<class Velocity, class Gamma>
+struct Push
+{
+    /* this is an optional extension for sub-sampling pushes that enables grid to particle interpolation
+     * for particle positions outside the super cell in one push
+     */
+    typedef typename pmacc::math::CT::make_Int<simDim,0>::type LowerMargin;
```
Comment (ax3l) [inline include/picongpu/particles/pusher/particlePusherAcceleration.hpp]: please use `using` in new code

File: include/picongpu/param/pusher.param (hunk 8 above)
Comment (steindev) [reply include/picongpu/param/pusher.param]: I will do so, but in a new PR as this is already merged. Or can I push to this PR again?

File: include/picongpu/particles/pusher/particlePusherAcceleration.hpp (hunk 9 above)
Comment (steindev) [reply include/picongpu/particles/pusher/particlePusherAcceleration.hpp]: Since i don't need this definition I want to remove it anyway.

Comment (ax3l) [reply include/picongpu/particles/pusher/particlePusherAcceleration.hpp]: Nope, won't work. This is a public concept used for caching when applying this algorithm to a species.

File: include/picongpu/param/pusher.param (hunk 8 above)
Comment (ax3l) [reply include/picongpu/param/pusher.param]: No, just do a follow-up PR.

**Pull request merged**